File: tests/compatibility/node/test-sdk.mjs
import assert from 'node:assert/strict';
import OpenAI from 'openai';
import Anthropic from '@anthropic-ai/sdk';

const baseURL = process.env.HALRO_COMPAT_BASE_URL ?? 'http://127.0.0.1:18088/v1';
const apiKey = 'gw_sdk_compatibility';
const client = new OpenAI({ apiKey, baseURL, maxRetries: 0, timeout: 5_000 });
const anthropic = new Anthropic({ apiKey, baseURL: baseURL.replace(/\/v1$/, ''), maxRetries: 0, timeout: 5_000 });

const message = await anthropic.messages.create({ model: 'compat-chat', max_tokens: 32, messages: [{ role: 'user', content: 'ping' }] });
assert.equal(message.content[0].text, 'compat-ok');
assert.equal(message.usage.output_tokens, 2);
const messageStream = await anthropic.messages.create({ model: 'compat-chat', max_tokens: 32, messages: [{ role: 'user', content: 'ping' }], stream: true });
let anthropicText = '';
for await (const event of messageStream) if (event.type === 'content_block_delta' && event.delta.type === 'text_delta') anthropicText += event.delta.text;
assert.equal(anthropicText, 'compat-ok');

const completion = await client.chat.completions.create({
  model: 'compat-chat',
  messages: [{ role: 'user', content: 'ping' }],
});
assert.equal(completion.choices[0].message.content, 'compat-ok');
assert.equal(completion.usage.total_tokens, 5);

const response = await client.responses.create({ model: 'compat-chat', input: 'ping', store: false });
assert.equal(response.output_text, 'compat-ok');
assert.equal(response.usage.total_tokens, 5);
assert.equal(response.store, false);

const responseStream = await client.responses.create({ model: 'compat-chat', input: 'ping', store: false, stream: true });
let responseText = '';
let responseCompleted = false;
for await (const event of responseStream) {
  if (event.type === 'response.output_text.delta') responseText += event.delta;
  if (event.type === 'response.completed') responseCompleted = true;
}
assert.equal(responseText, 'compat-ok');
assert.equal(responseCompleted, true);

const embedding = await client.embeddings.create({ model: 'compat-embedding', input: 'ping' });
assert.deepEqual(embedding.data[0].embedding, [0.125, -0.25, 0.5]);
assert.equal(embedding.usage.total_tokens, 2);

const stream = await client.chat.completions.create({
  model: 'compat-chat',
  messages: [{ role: 'user', content: 'ping' }],
  stream: true,
  stream_options: { include_usage: true },
});
let content = '';
let argumentsJSON = '';
let totalTokens;
for await (const chunk of stream) {
  if (chunk.usage) totalTokens = chunk.usage.total_tokens;
  for (const choice of chunk.choices) {
    content += choice.delta.content ?? '';
    for (const call of choice.delta.tool_calls ?? []) {
      argumentsJSON += call.function?.arguments ?? '';
    }
  }
}
assert.equal(content, 'compat-ok');
assert.deepEqual(JSON.parse(argumentsJSON), { value: 'ok' });
assert.equal(totalTokens, 7);

for (const [model, status] of [['error-budget', 403], ['error-rate', 429], ['error-provider', 502]]) {
  await assert.rejects(
    client.chat.completions.create({ model, messages: [{ role: 'user', content: 'ping' }] }),
    (error) => error instanceof OpenAI.APIError && error.status === status,
  );
}

File: tests/compatibility/python/test_sdk.py
import json
import os
import time
import urllib.request

import openai
import anthropic


BASE_URL = os.environ.get("HEIMDALL_COMPAT_BASE_URL", "http://127.0.0.1:18088/v1")
API_KEY = "gw_sdk_compatibility"


def client(api_key: str = API_KEY) -> openai.OpenAI:
    return openai.OpenAI(api_key=api_key, base_url=BASE_URL, max_retries=0, timeout=5.0)


def test_anthropic_messages() -> None:
    sdk = anthropic.Anthropic(api_key=API_KEY, base_url=BASE_URL.removesuffix("/v1"), max_retries=0, timeout=5.0)
    message = sdk.messages.create(model="compat-chat", max_tokens=32, messages=[{"role": "user", "content": "ping"}])
    assert message.content[0].text == "compat-ok"
    assert message.usage.output_tokens == 2
    text = ""
    with sdk.messages.stream(model="compat-chat", max_tokens=32, messages=[{"role": "user", "content": "ping"}]) as stream:
        for value in stream.text_stream:
            text += value
    assert text == "compat-ok"


def test_non_stream_and_embeddings() -> None:
    completion = client().chat.completions.create(
        model="compat-chat", messages=[{"role": "user", "content": "ping"}]
    )
    assert completion.choices[0].message.content == "compat-ok"
    assert completion.usage is not None and completion.usage.total_tokens == 5

    embedding = client().embeddings.create(model="compat-embedding", input="ping")
    assert embedding.data[0].embedding == [0.125, -0.25, 0.5]
    assert embedding.usage.total_tokens == 2

    response = client().responses.create(model="compat-chat", input="ping", store=False)
    assert response.output_text == "compat-ok"
    assert response.usage is not None and response.usage.total_tokens == 5
    assert response.store is False


def test_responses_stream() -> None:
    stream = client().responses.create(model="compat-chat", input="ping", store=False, stream=True)
    content = ""
    completed = False
    for event in stream:
        if event.type == "response.output_text.delta":
            content += event.delta
        if event.type == "response.completed":
            completed = True
    assert content == "compat-ok"
    assert completed


def test_stream_tool_calls_and_usage_only_chunk() -> None:
    stream = client().chat.completions.create(
        model="compat-chat",
        messages=[{"role": "user", "content": "ping"}],
        stream=True,
        stream_options={"include_usage": True},
    )
    content = ""
    arguments = ""
    total_tokens = None
    for chunk in stream:
        if chunk.usage is not None:
            total_tokens = chunk.usage.total_tokens
        for choice in chunk.choices:
            content += choice.delta.content or ""
            for call in choice.delta.tool_calls or []:
                arguments += call.function.arguments or ""
    assert content == "compat-ok"
    assert json.loads(arguments) == {"value": "ok"}
    assert total_tokens == 7


def test_error_matrix() -> None:
    cases = [("error-budget", 403), ("error-rate", 429), ("error-provider", 502)]
    for model, status in cases:
        try:
            client().chat.completions.create(
                model=model, messages=[{"role": "user", "content": "ping"}]
            )
            raise AssertionError(f"{model} unexpectedly succeeded")
        except openai.APIStatusError as exc:
            assert exc.status_code == status
    try:
        client("gw_invalid").chat.completions.create(
            model="compat-chat", messages=[{"role": "user", "content": "ping"}]
        )
        raise AssertionError("invalid key unexpectedly succeeded")
    except openai.AuthenticationError as exc:
        assert exc.status_code == 401


def test_disconnecting_stream_releases_server_resources() -> None:
    stream = client().chat.completions.create(
        model="slow-stream",
        messages=[{"role": "user", "content": "ping"}],
        stream=True,
    )
    next(iter(stream))
    stream.close()
    stats_url = BASE_URL.removesuffix("/v1") + "/compat/stats"
    deadline = time.monotonic() + 3
    while time.monotonic() < deadline:
        with urllib.request.urlopen(stats_url, timeout=1) as response:
            stats = json.load(response)
        if stats["active_streams"] == 0 and stats["canceled_streams"] >= 1:
            return
        time.sleep(0.05)
    raise AssertionError(f"stream resources were not released: {stats}")


if __name__ == "__main__":
    # Official Anthropic SDKs exercise x-api-key and anthropic-version headers.
    test_anthropic_messages()
    test_non_stream_and_embeddings()
    # Keep the executable CI entry point aligned with pytest discovery.
    test_responses_stream()
    test_stream_tool_calls_and_usage_only_chunk()
    test_error_matrix()
    test_disconnecting_stream_releases_server_resources()
